akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -627,6 +628,86 @@ final class Broadcast[T](val outputPorts: Int, val eagerCancel: Boolean) extends
 
 }
 
+object WireTap {
+  private val singleton = new WireTap[Nothing]
+
+  /**
+   * @see [[WireTap]]
+   */
+  def apply[T](): WireTap[T] = singleton.asInstanceOf[WireTap[T]]
+}
+
+/**
+ * Fan-out the stream to two output streams - a 'main' and a 'tap' one. Each incoming element is emitted
+ * to the 'main' output; elements are also emitted to the 'tap' output if there is demand;
+ * otherwise they are dropped.
+ *
+ * '''Emits when''' element is available and demand exists from the 'main' output; the element will
+ * also be sent to the 'tap' output if there is demand.
+ *
+ * '''Backpressures when''' the 'main' output backpressures
+ *
+ * '''Completes when''' upstream completes
+ *
+ * '''Cancels when''' the 'main' output cancels
+ *
+ */
+@InternalApi
+private[stream] final class WireTap[T] extends GraphStage[FanOutShape2[T, T, T]] {
+  val in: Inlet[T] = Inlet[T]("WireTap.in")
+  val outMain: Outlet[T] = Outlet[T]("WireTap.outMain")
+  val outTap: Outlet[T] = Outlet[T]("WireTap.outTap")
+  override def initialAttributes = DefaultAttributes.wireTap
+  override val shape: FanOutShape2[T, T, T] = new FanOutShape2(in, outMain, outTap)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var pendingTap: Option[T] = None
```

Comment:
As I think I said before, use an `akka.util.OptionVal` for this instead

Reply:
Ah yes, but, see: https://github.com/akka/akka/pull/23824#discussion_r146344829; `OptionVal`, as far as I can tell, only works for nullable types; but flows can of course be of non-nullable elements.

Any thoughts?